lib/src/convert_es6_proxy.dart

```diff
@@ -0,0 +1,59 @@
+@HtmlImport('convert_es6_proxy.html')
+library polymer_interop.lib.src.convert_es_proxy;
+
+import 'package:web_components/web_components.dart' show HtmlImport;
+
+import 'dart:js';
+
+import 'package:polymer_interop/polymer_interop.dart';
+export 'package:polymer_interop/polymer_interop.dart'
+    show convertToDart, convertToJs;
+
+final JsObject _polymerInteropDartES6 = context['Polymer']['PolymerInterop']['ES6'];
+
+final JsFunction createES6JsProxyForArray = _initES6ListProxySupport();
+
+final JsObject _Unsupported = _polymerInteropDartES6['Unsupported'];
+
+/// Hooks for ES6 Proxies
+///
+/// Hooks for getting and setting properties and methods from and to dart objects.
+///
+JsFunction _initES6ListProxySupport() {
+  <String, Function>{
+    '_dartArrayGet': (List instance, index) =>
+        (index is num) ? convertToJs(instance[index]) : _Unsupported,
+    '_dartArrayPut': (List instance, int index, val) {
+      if (index is num) {
+        instance[index] = convertToDart(val);
+      } else {
+        return _Unsupported;
+      }
+    },
+    '_dartArrayLength': (List instance) => instance.length,
+    '_dartArraySplice': (List instance, int index, int howmany, List items) {
```

Comment:
Will want to add some good tests for this one as it is easy to get subtly wrong ;)